### PR TITLE
[GLUTEN-3839][CH] Extend nested column pruning in vanilla spark

### PR DESCRIPTION
**Pull request by taiyang-li**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

(Fixes: \#3839)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Comment by taiyang-li** [comment]
Performance comparison: 

``` sql 
CREATE TEMPORARY VIEW test_table
USING org.apache.spark.sql.parquet
OPTIONS (
  path  "/data1/liyang/cppproject/spark/spark-3.3.2-bin-hadoop3/bigo_live_user_event"
) ;

select 
case when event.log_extra['tab_type'] in (5) then '1' else '0' end as entrance
from test_table
lateral view explode(events)  as event
where  event.log_extra['action'] in (13)  

set spark.gluten.sql.extendedGeneratorNestedColumnAliasing = true;
No rows selected (0.546 seconds)

set spark.gluten.sql.extendedGeneratorNestedColumnAliasing = false;
No rows selected (9.326 seconds)
```

**Comment by taiyang-li** [comment]
Another performance comparison on production.

Query: d_12768_1.sql 


Run query with `set spark.gluten.sql.extendedGeneratorNestedColumnAliasing = true;`
image
```
SubstraitFileSourceStep (read local files)
                                          Header: uid Nullable(Int64)
                                                  country Nullable(String)
                                                  events Nullable(Array(Nullable(Tuple(event_id Nullable(String), log_extra Nullable(Map(String, Nullable(String))), event_info Nullable(Map(String, Nullable(String)))))))
                                                  day Nullable(String)
```



Run query with `set spark.gluten.sql.extendedGeneratorNestedColumnAliasing = false`
image

```
SubstraitFileSourceStep (read local files)
                                              Header: uid Nullable(Int64)
                                                      country Nullable(String)
                                                      events Nullable(Array(Nullable(Tuple(time Nullable(Int64), lng Nullable(Int64), lat Nullable(Int64), net Nullable(String), event_id Nullable(String), log_extra Nullable(Map(String, Nullable(String))), event_info Nullable(Map(String, Nullable(String)))))))
                                                      day Nullable(String)
```

File: shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala [hunk 1]
```diff
@@ -1979,10 +1979,10 @@ object GlutenConfig {
       .booleanConf
       .createWithDefault(false)
 
-  val ENABLE_EXTENDED_GENERATOR_NESTED_COLUMN_ALIASING =
-    buildConf("spark.gluten.sql.extendedGeneratorNestedColumnAliasing")
+  val ENABLE_EXTENDED_COLUMN_PRUNING =
+    buildConf("spark.gluten.sql.extendedColumnPruning")
```
Comment (exmy) [inline shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala:1983]: `spark.gluten.sql.extendedColumnPruning.enabled` may be better.

File: backends-clickhouse/src/test/scala/org/apache/gluten/execution/hive/GlutenClickHouseHiveTableSuite.scala [hunk 2]
```diff
@@ -1490,15 +1490,92 @@ class GlutenClickHouseHiveTableSuite
...
+                  |LATERAL VIEW EXPLODE(events) AS event
+                  |LATERAL VIEW EXPLODE(split(event.log_extra['abflags_v3'], ',')) AS abflag
+                  |WHERE
+                  |event.event_id = 'xx'
+                  |AND event.event_info['dispatch_id'] IS NOT NULL
+                  |AND event.event_info['dispatch_id'] != ''
+                  |AND event.log_extra['scene'] = 'xxx'
+                  |LIMIT 100;
+      """.stripMargin)
+    assertFieldSizeAfterPruning(df3, 3)
```
Comment (exmy) [inline backends-clickhouse/src/test/scala/org/apache/gluten/execution/hive/GlutenClickHouseHiveTableSuite.scala:1574]: Missing compare result with vanilla spark.

File: backends-clickhouse/src/main/scala/org/apache/gluten/extension/ExtendedColumnPruning.scala [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+          case (attr, _) =>
+            attr.references.subsetOf(generatorOutputSet)
+        }
+
+      val pushedThrough = rewritePlanWithAliases(pj, attrToExtractValuesNotOnGenerator)
+
+      // We cannot push through if the child of generator is `MapType`.
+      g.generator.children.head.dataType match {
+        case _: MapType => return Some(pushedThrough)
+        case ArrayType(_: ArrayType, _) => return Some(pushedThrough)
```
[the remark this answers is not in the document]
Comment (taiyang-li) [reply backends-clickhouse/src/main/scala/org/apache/gluten/extension/ExtendedColumnPruning.scala:59]: Consider `lateral view explode(items.x)`, and the type of items is `array<struct<x map<string, string>>>`, then the type of items.x is `array<map<string, string>>`.  And we need to apply column pruning on generator expression. 

It is right also when `g.generator.children.head.dataType` is  `array<array>`. I had removed it from code.

File: backends-clickhouse/src/test/scala/org/apache/gluten/execution/hive/GlutenClickHouseHiveTableSuite.scala (hunk 2 above)
Comment (taiyang-li) [reply backends-clickhouse/src/test/scala/org/apache/gluten/execution/hive/GlutenClickHouseHiveTableSuite.scala]: fixed

**Review by exmy — APPROVED** [review]

**Pull request merged**